transitions/extensions/locking.py

```diff
@@ -3,36 +3,64 @@
 from threading import RLock
 import inspect
 
+try:
+    from contextlib import nested  # Python 2
+except ImportError:
+    from contextlib import ExitStack, contextmanager
+
+    @contextmanager
+    def nested(*contexts):
+        """
+        Reimplementation of nested in python 3.
+        """
+        with ExitStack() as stack:
+            for ctx in contexts:
+                stack.enter_context(ctx)
+            yield contexts
+
 
 class LockedMethod:
 
-    def __init__(self, lock, func):
-        self.lock = lock
+    def __init__(self, managers, func):
+        self.managers = managers
         self.func = func
 
     def __call__(self, *args, **kwargs):
-        with self.lock:
+        with nested(*self.managers):
             return self.func(*args, **kwargs)
 
 
 class LockedEvent(Event):
 
     def trigger(self, model, *args, **kwargs):
-        with self.machine.rlock:
+        with nested(*self.machine.managers):
             return super(LockedEvent, self).trigger(model, *args, **kwargs)
 
 
 class LockedMachine(Machine):
 
     def __init__(self, *args, **kwargs):
-        self.rlock = RLock()
+        self.managers = []
+
+        try:
+            self.managers.append(kwargs.pop('lock'))
+        except KeyError:
+            pass
+
+        try:
+            self.managers.extend(kwargs.pop('managers'))
+        except KeyError:
+            pass
+
```

Comment:
Having two keywords which are pretty much handled the same way might be a bit confusing. I do not know which word might encompass both (`context` maybe?).